File: mekaneck/virtualbrain/operators/sentropy_ops.py
"""
S-Entropy Operators: NAVIGATE, GRAD_S, UPDATE_SK, UPDATE_ST, UPDATE_SE.

Implements S-entropy space navigation and update operations.
"""

from typing import List, Callable, Tuple
import numpy as np

from ..core.types import SCoord
from ..core.constants import TAU_CONSCIOUSNESS


def NAVIGATE(
    current: SCoord,
    target: SCoord,
    step_size: float = 0.1,
    max_steps: int = 100,
) -> List[SCoord]:
    """
    NAVIGATE operator: Find path from current to target in S-space.

    Uses gradient descent navigation with O(log n) complexity
    via ternary trisection.

    Args:
        current: Starting S-entropy coordinate
        target: Target S-entropy coordinate
        step_size: Step size for each iteration
        max_steps: Maximum navigation steps

    Returns:
        Path of SCoord from current to target
    """
    path = [current]
    state = current

    for _ in range(max_steps):
        distance = state.distance(target)

        if distance < step_size:
            path.append(target)
            break

        # Compute gradient direction
        grad = state.gradient(target)

        # Update state
        state = state.update(
            delta_sk=step_size * grad[0],
            delta_st=step_size * grad[1],
            delta_se=step_size * grad[2],
        )
        path.append(state)

    return path


def GRAD_S(
    s_coord: SCoord,
    field: Callable[[SCoord], float],
    epsilon: float = 1e-6,
) -> np.ndarray:
    """
    GRAD_S operator: Compute gradient of field in S-space.

    Uses central differences for numerical gradient.

    Args:
        s_coord: Point at which to compute gradient
        field: Scalar field function f(SCoord) -> float
        epsilon: Step size for finite differences

    Returns:
        Gradient vector [df/dSk, df/dSt, df/dSe]
    """
    grad = np.zeros(3)

    # Gradient in Sk direction
    s_plus = s_coord.update(delta_sk=epsilon)
    s_minus = s_coord.update(delta_sk=-epsilon)
    grad[0] = (field(s_plus) - field(s_minus)) / (2 * epsilon)

    # Gradient in St direction
    s_plus = s_coord.update(delta_st=epsilon)
    s_minus = s_coord.update(delta_st=-epsilon)
    grad[1] = (field(s_plus) - field(s_minus)) / (2 * epsilon)

    # Gradient in Se direction
    s_plus = s_coord.update(delta_se=epsilon)
    s_minus = s_coord.update(delta_se=-epsilon)
    grad[2] = (field(s_plus) - field(s_minus)) / (2 * epsilon)

    return grad


def UPDATE_SK(
    s_coord: SCoord,
    information_gain: float,
) -> SCoord:
    """
    UPDATE_SK operator: Update knowledge entropy from observation.

    Sk_new = Sk_old - I(observation)

    Information reduces uncertainty (entropy).

    Args:
        s_coord: Current S-entropy coordinate
        information_gain: Information gained (in normalized units)

    Returns:
        Updated SCoord with reduced Sk
    """
    return s_coord.update(delta_sk=-information_gain)


def UPDATE_ST(
    s_coord: SCoord,
    tau_circuit: float,
    tau_max: float = TAU_CONSCIOUSNESS,
) -> SCoord:
    """
    UPDATE_ST operator: Update temporal entropy from circuit completion.

    St_new = St_old + tau_circuit / tau_max

    Circuit completion increases temporal distance.

    Args:
        s_coord: Current S-entropy coordinate
        tau_circuit: Circuit completion duration
        tau_max: Maximum circuit time (normalization)

    Returns:
        Updated SCoord with increased St
    """
    delta_st = tau_circuit / tau_max
    return s_coord.update(delta_st=delta_st)


def UPDATE_SE(
    s_coord: SCoord,
    trajectory_step: float,
    dt: float = 0.01,
) -> SCoord:
    """
    UPDATE_SE operator: Update evolution entropy from trajectory step.

    Se_new = Se_old + |dS/dt| * dt

    Trajectory progression increases evolution entropy.

    Args:
        s_coord: Current S-entropy coordinate
        trajectory_step: Magnitude of trajectory step |dS|
        dt: Time step

    Returns:
        Updated SCoord with increased Se
    """
    delta_se = trajectory_step * dt
    return s_coord.update(delta_se=delta_se)


def entropy_field(s_coord: SCoord) -> float:
    """
    Compute total entropy field value at coordinate.

    S_total = Sk + St + Se

    Args:
        s_coord: S-entropy coordinate

    Returns:
        Total entropy (sum of components)
    """
    return s_coord.sk + s_coord.st + s_coord.se


def free_energy_field(
    s_coord: SCoord,
    temperature: float = 300.0,
    internal_energy: float = 0.0,
) -> float:
    """
    Compute Helmholtz free energy F = U - TS.

    Args:
        s_coord: S-entropy coordinate
        temperature: Temperature in Kelvin
        internal_energy: Internal energy U

    Returns:
        Free energy F
    """
    from ..core.constants import BOLTZMANN_CONSTANT

    S = s_coord.entropy_magnitude() * BOLTZMANN_CONSTANT
    return internal_energy - temperature * S


def minimize_free_energy(
    initial: SCoord,
    temperature: float = 300.0,
    max_iterations: int = 100,
    learning_rate: float = 0.1,
) -> Tuple[SCoord, List[float]]:
    """
    Find S-coordinate minimizing free energy.

    Uses gradient descent on the free energy landscape.

    Args:
        initial: Starting coordinate
        temperature: Temperature for free energy
        max_iterations: Maximum iterations
        learning_rate: Step size

    Returns:
        (final_coord, energy_history)
    """

    def F(s):
        return free_energy_field(s, temperature)

    current = initial
    history = [F(current)]

    for _ in range(max_iterations):
        grad = GRAD_S(current, F)
        current = current.update(
            delta_sk=-learning_rate * grad[0],
            delta_st=-learning_rate * grad[1],
            delta_se=-learning_rate * grad[2],
        )
        history.append(F(current))

        # Check convergence
        if len(history) > 1 and abs(history[-1] - history[-2]) < 1e-8:
            break

    return current, history

File: virtual_brain/crates/vb-operators/src/sentropy_ops.rs
//! S-Entropy Operators: NAVIGATE, GRAD_S, UPDATE_*.
//!
//! Operators for navigating S-entropy space.

use vb_core::types::SCoord;

const EPSILON: f64 = 1e-10;

/// NAVIGATE operator: Gradient descent path from current to target.
pub fn navigate(
    current: &SCoord,
    target: &SCoord,
    step_size: f64,
    max_steps: usize,
) -> Vec<SCoord> {
    let mut path = vec![*current];
    let mut pos = *current;

    for _ in 0..max_steps {
        let dist = pos.distance(target);
        if dist < EPSILON {
            break;
        }

        let (gsk, gst, gse) = pos.gradient(target);
        let step = step_size.min(dist);

        pos = pos.update(gsk * step, gst * step, gse * step);
        path.push(pos);
    }

    path
}

/// GRAD_S operator: Numerical gradient of a field.
pub fn grad_s<F>(s: &SCoord, field: F, epsilon: f64) -> [f64; 3]
where
    F: Fn(&SCoord) -> f64,
{
    let f0 = field(s);

    // Partial derivatives
    let dsk = if s.sk + epsilon <= 1.0 {
        let s_plus = SCoord::new_unchecked(s.sk + epsilon, s.st, s.se);
        (field(&s_plus) - f0) / epsilon
    } else if s.sk - epsilon >= 0.0 {
        let s_minus = SCoord::new_unchecked(s.sk - epsilon, s.st, s.se);
        (f0 - field(&s_minus)) / epsilon
    } else {
        0.0
    };

    let dst = if s.st + epsilon <= 1.0 {
        let s_plus = SCoord::new_unchecked(s.sk, s.st + epsilon, s.se);
        (field(&s_plus) - f0) / epsilon
    } else if s.st - epsilon >= 0.0 {
        let s_minus = SCoord::new_unchecked(s.sk, s.st - epsilon, s.se);
        (f0 - field(&s_minus)) / epsilon
    } else {
        0.0
    };

    let dse = if s.se + epsilon <= 1.0 {
        let s_plus = SCoord::new_unchecked(s.sk, s.st, s.se + epsilon);
        (field(&s_plus) - f0) / epsilon
    } else if s.se - epsilon >= 0.0 {
        let s_minus = SCoord::new_unchecked(s.sk, s.st, s.se - epsilon);
        (f0 - field(&s_minus)) / epsilon
    } else {
        0.0
    };

    [dsk, dst, dse]
}

/// UPDATE_SK operator: Update knowledge entropy based on information gain.
pub fn update_sk(s: &SCoord, information_gain: f64) -> SCoord {
    // More information reduces knowledge entropy
    let delta_sk = -information_gain * s.sk;
    s.update(delta_sk, 0.0, 0.0)
}

/// UPDATE_ST operator: Update temporal entropy based on circuit timescale.
pub fn update_st(s: &SCoord, tau_circuit: f64, tau_max: f64) -> SCoord {
    // Longer timescale increases temporal entropy
    let delta_st = (tau_circuit / tau_max) * (1.0 - s.st);
    s.update(0.0, delta_st, 0.0)
}

/// UPDATE_SE operator: Update evolution entropy based on trajectory step.
pub fn update_se(s: &SCoord, trajectory_step: f64, dt: f64) -> SCoord {
    // Larger trajectory steps increase evolution entropy
    let delta_se = trajectory_step * dt * (1.0 - s.se);
    s.update(0.0, 0.0, delta_se)
}

/// Entropy field: Total entropy magnitude.
pub fn entropy_field(s: &SCoord) -> f64 {
    s.entropy_magnitude()
}

/// Free energy field: F = U - T*S
pub fn free_energy_field(s: &SCoord, temperature: f64, internal_energy: f64) -> f64 {
    internal_energy - temperature * s.entropy_magnitude()
}

/// Minimize free energy using gradient descent.
pub fn minimize_free_energy(
    initial: &SCoord,
    temperature: f64,
    max_iterations: usize,
    learning_rate: f64,
) -> SCoord {
    let mut s = *initial;

    for _ in 0..max_iterations {
        let grad = grad_s(&s, |coord| free_energy_field(coord, temperature, 1.0), 1e-6);

        // Gradient descent step
        let new_s = s.update(
            -learning_rate * grad[0],
            -learning_rate * grad[1],
            -learning_rate * grad[2],
        );

        // Check convergence
        if s.distance(&new_s) < EPSILON {
            break;
        }
        s = new_s;
    }

    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_navigate() {
        let start = SCoord::origin();
        let target = SCoord::new(0.5, 0.5, 0.5).unwrap();
        let path = navigate(&start, &target, 0.1, 100);

        assert!(path.len() > 1);
        assert!(path.last().unwrap().distance(&target) < 0.15);
    }

    #[test]
    fn test_grad_s() {
        let s = SCoord::new(0.5, 0.5, 0.5).unwrap();
        let grad = grad_s(&s, entropy_field, 1e-6);

        // Gradient should point toward increasing entropy
        assert!(grad[0] > 0.0);
        assert!(grad[1] > 0.0);
        assert!(grad[2] > 0.0);
    }

    #[test]
    fn test_update_sk() {
        let s = SCoord::new(0.8, 0.5, 0.5).unwrap();
        let updated = update_sk(&s, 0.5);
        assert!(updated.sk < s.sk);
    }

    #[test]
    fn test_update_st() {
        let s = SCoord::new(0.5, 0.3, 0.5).unwrap();
        let updated = update_st(&s, 0.5, 1.0);
        assert!(updated.st > s.st);
    }
}
